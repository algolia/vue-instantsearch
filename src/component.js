import suit from './suit.js';

export default {
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  created() {
    this.widgetFactory = this.connector(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdate = Object.keys(widgetParams).some(name =>
        widgetParams[name](previous[name], next[name])
      );

      if (shouldUpdate) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  methods: {
    suit(...args) {
      return suit(this.widgetName, ...args);
    },
    updateData(state = {}) {
      this.state = state;
    },
  },
};
