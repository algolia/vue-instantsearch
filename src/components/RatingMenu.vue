<template>
  <div
    v-if="state"
    :class="suit()"
  >
    <slot
      :items="state.items"
      :refine="state.refine"
    >
      <svg
        xmlns="http://www.w3.org/2000/svg"
        style="display: none;"
      >
        <symbol
          :id="suit('starSymbol')"
          viewBox="0 0 24 24"
        >
          <path d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z" />
        </symbol>
        <symbol
          :id="suit('starEmptySymbol')"
          viewBox="0 0 24 24"
        >
          <path d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z" />
        </symbol>
      </svg>

      <ul :class="suit('list')">
        <li
          v-for="(item, key) in state.items"
          :key="key"
          :class="[suit('item'), item.isRefined && suit('item', 'selected')]"
        >
          <a
            href="#"
            :aria-label="`${item.value} & Up`"
            :class="suit('link')"
            @click.prevent="toggleRefinement(item.value)"
          >
            <template v-for="n in max">
              <svg
                v-if="n <= item.value"
                aria-hidden="true"
                width="24"
                height="24"
                :class="[suit('starIcon'), suit('starIcon--full')]"
                :key="n"
              >
                <use :xlink:href="`#${suit('starSymbol')}`" />
              </svg>

              <svg
                v-else
                :class="[suit('starIcon'), suit('starIcon--empty')]"
                aria-hidden="true"
                width="24"
                height="24"
                :key="n"
              >
                <use :xlink:href="`#${suit('starEmptySymbol')}`" />
              </svg>
            </template>

            <span
              :class="suit('label')"
              aria-hidden="true"
            >
              &amp; Up
            </span>
            <span :class="suit('count')">
              {{ item.count }}
            </span>
          </a>
        </li>
      </ul>
    </slot>
  </div>
</template>

<script>
import { connectStarRating } from 'instantsearch.js/es/connectors';
import { createPanelConsumerMixin } from '../mixins/panel';
import { createWidgetMixin } from '../mixins/widget';

export default {
  mixins: [
    createWidgetMixin({ connector: connectStarRating }),
    createPanelConsumerMixin({
      mapStateToCanRefine: state => !state.hasNoResults,
    }),
  ],
  props: {
    attribute: {
      type: String,
      required: true,
    },
    min: {
      type: Number,
      default: 1,
    },
    max: {
      type: Number,
      default: 5,
    },
  },
  data() {
    return {
      widgetName: 'RatingMenu',
    };
  },
  computed: {
    widgetParams() {
      return {
        attributeName: this.attribute,
        min: this.min,
        max: this.max,
      };
    },
  },
  methods: {
    toggleRefinement(value) {
      this.state.refine(value);
    },
  },
};
</script>
