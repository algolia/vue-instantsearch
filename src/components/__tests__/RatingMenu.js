import { mount } from '@vue/test-utils';
import RatingMenu from '../RatingMenu.vue';
import { __setState } from '../../component';
jest.mock('../../component');

it('renders correctly', () => {
  __setState({
    items: [
      { isRefined: false, count: 20, value: 1 },
      { isRefined: false, count: 3, value: 2 },
      { isRefined: false, count: 2000, value: 3 },
      { isRefined: false, count: 8, value: 4 },
    ],
  });
  const wrapper = mount(RatingMenu, {
    propsData: {
      attribute: 'hi',
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly when refined', () => {
  __setState({
    items: [
      { isRefined: false, count: 20, value: 1 },
      { isRefined: false, count: 3, value: 2 },
      { isRefined: false, count: 2000, value: 3 },
      { isRefined: true, count: 8, value: 4 },
    ],
  });
  const wrapper = mount(RatingMenu, {
    propsData: {
      attribute: 'hi',
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

it('calls refine when clicked on link', () => {
  __setState({
    items: [
      { isRefined: false, count: 20, value: 1 },
      { isRefined: false, count: 3, value: 2 },
      { isRefined: false, count: 2000, value: 3 },
      { isRefined: false, count: 8, value: 4 },
    ],
    refine: jest.fn(),
  });
  const wrapper = mount(RatingMenu, {
    propsData: {
      attribute: 'hi',
    },
  });
  const link = wrapper.find('.ais-RatingMenu-link');
  link.trigger('click');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith(1);
});
