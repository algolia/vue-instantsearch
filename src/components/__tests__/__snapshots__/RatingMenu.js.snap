// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `

<div class="ais-RatingMenu">
  <svg xmlns="http://www.w3.org/2000/svg"
       style="display: none;"
  >
    <symbol id="ais-RatingMenu-starSymbol"
            viewbox="0 0 24 24"
    >
      <path d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z">
      </path>
    </symbol>
    <symbol id="ais-RatingMenu-starEmptySymbol"
            viewbox="0 0 24 24"
    >
      <path d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z">
      </path>
    </symbol>
  </svg>
  <ul class="ais-RatingMenu-list">
    <li class="ais-RatingMenu-item">
      <a href="#"
         aria-label="1 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          20
        </span>
      </a>
    </li>
    <li class="ais-RatingMenu-item">
      <a href="#"
         aria-label="2 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          3
        </span>
      </a>
    </li>
    <li class="ais-RatingMenu-item">
      <a href="#"
         aria-label="3 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          2000
        </span>
      </a>
    </li>
    <li class="ais-RatingMenu-item">
      <a href="#"
         aria-label="4 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          8
        </span>
      </a>
    </li>
  </ul>
</div>

`;

exports[`renders correctly when refined 1`] = `

<div class="ais-RatingMenu">
  <svg xmlns="http://www.w3.org/2000/svg"
       style="display: none;"
  >
    <symbol id="ais-RatingMenu-starSymbol"
            viewbox="0 0 24 24"
    >
      <path d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z">
      </path>
    </symbol>
    <symbol id="ais-RatingMenu-starEmptySymbol"
            viewbox="0 0 24 24"
    >
      <path d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z">
      </path>
    </symbol>
  </svg>
  <ul class="ais-RatingMenu-list">
    <li class="ais-RatingMenu-item">
      <a href="#"
         aria-label="1 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          20
        </span>
      </a>
    </li>
    <li class="ais-RatingMenu-item">
      <a href="#"
         aria-label="2 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          3
        </span>
      </a>
    </li>
    <li class="ais-RatingMenu-item">
      <a href="#"
         aria-label="3 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          2000
        </span>
      </a>
    </li>
    <li class="ais-RatingMenu-item ais-RatingMenu-item--selected">
      <a href="#"
         aria-label="4 &amp; Up"
         class="ais-RatingMenu-link"
      >
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--full"
        >
          <use xlink:href="#ais-RatingMenu-starSymbol">
          </use>
        </svg>
        <svg aria-hidden="true"
             width="24"
             height="24"
             class="ais-RatingMenu-starIcon ais-RatingMenu-starIcon--empty"
        >
          <use xlink:href="#ais-RatingMenu-starEmptySymbol">
          </use>
        </svg>
        <span aria-hidden="true"
              class="ais-RatingMenu-label"
        >
          &amp; Up
        </span>
        <span class="ais-RatingMenu-count">
          8
        </span>
      </a>
    </li>
  </ul>
</div>

`;
